### DISCUSSION TITLE
What is the best event for detecting when a player left/right clicks on either air/block/entity?

### DISCUSSION THREAD

**Discussion by Technerder** [Minestom/Minestom#934] [Help]
Similar to Spigot's `PlayerInteractEvent`

**Comment by ZakShearman** [comment]
You must use multiple events and then either make your own utility to handle them as one or handle them separately as you wish.

PlayerEntityInteractEvent - right click entity.
PlayerBlockInteractEvent - right-click block, also called when a block is placed.
PlayerUseItemOnBlockEvent - right click block with item (but not a block item)
PlayerUseItemEvent - right click air
PlayerHandAnimationEvent - I am pretty sure this is left-click, although I haven't used it myself. It will probably also trigger if done on a block.

I've used PlayerBlockInteractEvent, PlayerUseItemOnBlockEvent and PlayerUseItemEvent but I'm not sure of the function for PlayerHandAnimationEvent really. Javadocs don't say much about it

**Reply by ArvedEh** [reply]
For the PlayerHandAnimationEvent, maybe it gets sent when the player sends [a swing packet (0x06)](https://wiki.vg/Protocol#Entity_Animation_.28clientbound.29)?

Edit: Found it in the Minestom source: https://github.com/Minestom/Minestom/blob/master/src/main/java/net/minestom/server/listener/AnimationListener.java
